compiler/noirc_evaluator/src/ssa/opt/formal_verifications.rs

```diff
@@ -1,7 +1,473 @@
+use std::collections::HashMap;
+
+use crate::ssa::ir::dfg::DataFlowGraph;
+use crate::ssa::ir::function::{Function, FunctionId};
+use crate::ssa::ir::instruction::{Binary, FvInstruction, Instruction, InstructionId};
+use crate::ssa::ir::map::Id;
+use crate::ssa::ir::value::{Value, ValueId};
 use crate::ssa::ssa_gen::Ssa;
 
 impl Ssa {
     pub(crate) fn formal_verifications_optimization(mut self) -> Self {
+        // There is a loop bound to avoid infinite run.
+        for _ in 0..1000 {
+            if !self.has_call_instruction() {
+                break;
+            }
+            self.inline_first_call();
+        }
         self
     }
+
+    fn has_call_instruction(&self) -> bool {
+        let mut is_done = true;
+        for function in self.functions.values() {
+            let dfg = &function.dfg;
+            for fv_instruction in dfg.fv_instructions.iter() {
+                match fv_instruction {
+                    FvInstruction::Requires(instruction) => match instruction {
+                        Instruction::Call { func: _, arguments: _ } => {
+                            is_done = false;
+                        }
+                        _ => {}
+                    },
+                    FvInstruction::Ensures(instruction) => match instruction {
+                        Instruction::Call { func: _, arguments: _ } => {
+                            is_done = false;
+                        }
+                        _ => {}
+                    },
+                }
+            }
+        }
+        !is_done
+    }
+
+    pub(crate) fn inline_first_call(&mut self) {
+        // Maps a function and the new values that will be added to its dfg.values.
+        let mut update_func_with_values: Vec<(FunctionId, Vec<(usize, Value)>)> = Vec::new();
+
+        // Maps a function and the new fv instructions which will overwrite the old ones.
+        let mut update_func_fv_instructions: Vec<(FunctionId, Vec<FvInstruction>)> = Vec::new();
+
+        // For each function that will be modified we have to right shift the fv instructions which will not be inlined
+        // so we could make "space" for the ones which will be added by the inlining.
+        let mut update_func_fv_right_shift: Vec<(FunctionId, Vec<(usize, usize)>)> = Vec::new();
+
+        // We also have to update the results map, because we need to update the indexes which
+        let mut update_func_dfg_results: Vec<(FunctionId, Vec<(InstructionId, Vec<ValueId>)>)> =
+            Vec::new();
```

Comment:
Done for the second suggestions. I didn't want to change the structure from Vector to HashMap because we are not indexing the collection anywhere. We are only doing linear iterations over it. Therefore we would not improve our 
asymptotic computational complexity if we did with HashMap